lcls_tools/common/data_analysis/projection_fit/projection_fit.py

```diff
@@ -0,0 +1,130 @@
+    and if prior assumptions (bayesian regression) about the data should be used when performing the fit.
+        Additionally there is an option to visualize the fitted data and priors.
+    -To perform a 1d fit, call fit_projection(projection_data={*data_to_fit*})
+    ------------------------
+    Arguments:
+    model: MethodBase (this argument is a child class object of method base
+        e.g GaussianModel & DoubleGaussianModel)
+    visualize_priors: bool (shows plots of the priors and init guess distribution before fit)
+    use_priors: bool (incorporates prior distribution information into fit)
+    visualize_fit: bool (visualize the parameters as a function of the forward function
+        from our model compared to distribution data)
+    """
+    # come up with better name
+    model_config = ConfigDict(arbitrary_types_allowed=True)
+    model: MethodBase
+    visualize_priors: bool = False
+    use_priors: bool = False
+    visualize_fit: bool = False
+
+    def normalize(self, old_data: np.ndarray) -> np.ndarray:
+        """
+        Normalize a 1d array by scaling and shifting data
+        s.t. data is between 0 and 1
+        """
+        data = old_data.copy()
+        normalized_data = data / (np.max(data))
+        return normalized_data
+
+    def unnormalize_model_params(
+        self, params_dict: dict, projection_data: np.ndarray
+    ) -> np.ndarray:
+        """
+        Takes fitted and normalized params and returns them
+        to unnormalized values i.e the true fitted values of the distribution
+        """
+        max_value = np.max(projection_data)
+        length = len(projection_data)
+        for key, val in params_dict.items():
+            if "sigma" in key or "mean" in key:
+                true_fitted_val = val * length
+            else:
+                true_fitted_val = val * max_value
+            temp = {key: true_fitted_val}
+            params_dict.update(temp)
+        return params_dict
+
+    def model_setup(self, projection_data=np.ndarray) -> None:
+        """sets up the model and plots init_values/priors"""
+        self.model.profile_data = projection_data
+        if self.visualize_priors:
+            self.model.plot_init_values()
+            self.model.plot_priors()
+
+
+    def fit_model(self) -> scipy.optimize._optimize.OptimizeResult:
+        """
+        Fits model params to distribution data and plots the fitted params
+        as a function of the model.
+        Returns optimizeResult object
+        """
+        x = np.linspace(0, 1, len(self.model.profile_data))
+        y = self.model.profile_data
+
+
+        #TODO:self.model_init_values is now a dictionary -> change to list for minimize
+        #making an ordered list here is making this fit_model model dependent since it needs knowledge of the params list ordering
+
+        res = scipy.optimize.minimize(
+            self.model.loss,
```

Comment:
I think a see a potential problem with making forward and log prior receive a dictionary. Model.loss calls forward  and log prior. The problem is that it is scipy.optimize.minimize would pass an unpacked list of length == len(model.init_values) to model.loss on reiterations.
I can't think of how I would be able to resolve this.

Reply:
change the calls in model.loss to model._forward and model._log_prior which take np arrays instead. The public versions can take dicts as arguments